buildsystem/generator/golang.go

```diff
@@ -0,0 +1,597 @@
+	return catcher.Resolve()
+}
+
+// GolangVariant defines a mapping between distros that run packages and the
+// golang packages to run.
+type GolangVariant struct {
+	Name    string   `yaml:"name"`
+	Distros []string `yaml:"distros"`
+	// Packages lists the names or paths of the package relative to the root
+	// package, which can be one of the following:
+	// - A name of a package explicitly listed in the packages section.
+	// - A path of a package explicitly listed in the packages section.
+	// - A path of a package that matches a path to a directory containing tests
+	//   relative to the root package.
+	// - A package tag, beginning with a period (e.g. ".tag").
+	Packages []VariantPackage `yaml:"packages"`
+	// Options are variant-specific options that modify test execution.
+	// Explicitly setting these values will override options specified under the
+	// definitions of packages.
+	Options *GolangRuntimeOptions `yaml:"options,omitempty"`
+}
+
+func (gv *GolangVariant) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(gv.Name == "", "missing variant name")
+	catcher.NewWhen(len(gv.Distros) == 0, "need to specify at least one distro")
+	catcher.NewWhen(len(gv.Packages) == 0, "need to specify at least one package")
+	pkgPaths := map[string]struct{}{}
+	pkgNames := map[string]struct{}{}
+	pkgTags := map[string]struct{}{}
+	for _, pkg := range gv.Packages {
+		catcher.Wrap(pkg.Validate(), "invalid package reference")
+		if path := pkg.Path; path != "" {
+			if _, ok := pkgPaths[path]; ok {
+				catcher.Errorf("duplicate reference to package path '%s'", path)
+			}
+			pkgPaths[path] = struct{}{}
+		}
+		if name := pkg.Name; name != "" {
+			if _, ok := pkgNames[name]; ok {
+				catcher.Errorf("duplicate reference to package name '%s'", name)
+			}
+			pkgNames[name] = struct{}{}
+		}
+		if tag := pkg.Tag; tag != "" {
+			if _, ok := pkgTags[tag]; ok {
+				catcher.Errorf("duplicate reference to package tag '%s'", tag)
+			}
+			pkgTags[tag] = struct{}{}
+		}
+	}
+	if gv.Options != nil {
+		catcher.Wrap(gv.Options.Validate(), "invalid runtime options")
+	}
+	return catcher.Resolve()
+}
+
+type VariantPackage struct {
+	Name string `yaml:"name,omitempty"`
+	Path string `yaml:"path,omitempty"`
+	Tag  string `yaml:"tag,omitempty"`
+}
+
+func (vp *VariantPackage) Validate() error {
+	var numRefs int
+	for _, ref := range []string{vp.Name, vp.Path, vp.Tag} {
+		if ref != "" {
+			numRefs++
+		}
+	}
+	if numRefs != 1 {
+		return errors.New("must specify exactly one of the following: name, path, or tag")
+	}
+	return nil
+}
+
+// GolangRuntimeOptions specify additional options to modify behavior of runtime
+// execution.
+type GolangRuntimeOptions struct {
+	// TODO: add more go test flags? It might be somewhat onerous to maintain
+	// all the possible go test flags here, but parsing the individual flags to
+	// merge them is also painful.
+	Args []string `yaml:"args,omitempty"`
+}
+
+// TODO: validate options duplication.
```

Comment:
yes this would be cool

Reply:
Cool, but would be ugly and I don't think it's worth it to deal with for now since it would mostly devolve into string parsing/checking/deduplication and isn't critical for the config generation right now.